### PR TITLE
[cxx-interop] Implements constructor for std::span from UnsafeBufferPointer

### PR DESCRIPTION
**Pull request by susmonteiro**
Includes tests for:
- init `std::span` from UnsafeBufferPointer
- for loop
- map
- filter
- init swift `Array` from `std::span`
- `std::span` of strings

### CODE REVIEW

**Comment by susmonteiro** [comment]
@swift-ci please smoke test

**Comment by susmonteiro** [comment]
@swift-ci please smoke test

**Review by egorzhdan — COMMENTED** [review]
Looks awesome!
I have a couple of minor comments + could you please reformat the new code with clang-format?

File: lib/ClangImporter/ClangDerivedConformances.cpp [hunk 1]
```diff
@@ -1116,16 +1130,103 @@ void swift::conformToCxxFunctionIfNeeded(
 
   auto constructExpr =
       dyn_cast_or_null<clang::CXXConstructExpr>(bindTempExpr->getSubExpr());
+
```
[the remark this answers is not in the document]
Comment (susmonteiro) [reply lib/ClangImporter/ClangDerivedConformances.cpp:1133]: Yes sorry!

File: stdlib/public/Cxx/CxxSpan.swift [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  associatedtype Size: BinaryInteger
+
+  init()
+  init(_ unsafePointer : UnsafePointer<Element>, _ count: Size)
+}
+
+extension CxxSpan {
+  /// Creates a C++ span from a Swift UnsafeBufferPointer
+  @inlinable
+  public init(_ unsafeBufferPointer: __shared UnsafeBufferPointer<Element>) {
```
[the remark this answers is not in the document]
Comment (susmonteiro) [reply stdlib/public/Cxx/CxxSpan.swift]: Thanks. Fixed

File: stdlib/public/Cxx/CxxSpan.swift [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// A C++ type that an object that can refer to a contiguous sequence of objects .
```
Comment (egorzhdan) [inline stdlib/public/Cxx/CxxSpan.swift:13]: I think a word is missing in this sentence, e.g. "A C++ type that _is_ an object that can refer". Also, there is a space before period 😉

**Review by egorzhdan — COMMENTED** [review]
Just one tiny comment left about the documentation comment. Otherwise LGTM!

File: stdlib/public/Cxx/CxxSpan.swift (hunk 3 above)
Comment (susmonteiro) [reply stdlib/public/Cxx/CxxSpan.swift]: Thanks! Fixed

**Comment by susmonteiro** [comment]
@swift-ci please smoke test

**Review by egorzhdan — APPROVED** [review]
Awesome, LGTM!

**Pull request merged**